docs/_tooling/extensions/score_source_code_linker/parse_source_files.py

```diff
@@ -0,0 +1,123 @@
+import subprocess
+import os
+import argparse
+import collections
+import functools
+import json
+from typing import Optional
+from pathlib import Path
+
+
+TAGS = [
+    "# req-traceability:",
+    "# req-Id:",
+]
+
+GITHUB_BASE_URL = "https://github.com/eclipse-score/score/blob/"
+
+
+def extract_requirements(
+    source_file: str, git_hash_func: Optional[callable] = None
+) -> dict[str, list]:
+    """
+    This extracts the file-path, lineNr as well as the git hash of the file where a tag was found.
+
+    Args:
+        source_file (str): path to source file that should be parsed.
+        git_hash_func (Optional[callable]): Optional parameter only supplied during testing. If left empty func 'get_git_hash' is used.
+
+    Returns:
+        # TODO: change these links
+        Returns dictionary per file like this:
+        {
+            "TOOL_REQ__toolchain_sphinx_needs_build__requirement_linkage_types": [
+                    https://github.com/eclipse-score/score/blob/3b3397ebc2777f47b1ae5258afc4d738095adb83/_tooling/extensions/score_metamodel/utils.py,
+                    ... # further found places of the same ID if there are any
+                ]
+            "TOOL_REQ__toolchain_sphinx_needs_build__...": [
+                    https://github.com/eclipse-score/score/blob/3b3397ebc2777f47b1ae5258afc4d738095adb83/_tooling/extensions/score_metamodel/checks/id.py,
+                    ... # places where this ID as found
+            ]
+        }
+    """
+    requirement_mapping = collections.defaultdict(list)
+    get_hash = git_hash_func if git_hash_func is not None else get_git_hash
+    with open(source_file, "r") as f:
+        for line_number, line in enumerate(f):
+            line_number = line_number + 1
+            line = line.strip()
+            if any(x in line for x in TAGS):
+                hash = get_hash(source_file)
+                cleaned_line = (
+                    line.replace("'", "").replace('"', "").replace(",", "").strip()
+                )
+                check_tag = cleaned_line.split(":")[1].strip()
+                if check_tag:
+                    req_id = cleaned_line.split(":")[-1].strip()
+                    link = f"{GITHUB_BASE_URL}{hash}/{source_file}#L{line_number}"
+                    requirement_mapping[req_id].append(link)
+    return requirement_mapping
+
+
+def get_git_hash(file_path: str) -> str:
+    """
+    Grabs the latest git hash found for perticular file
+
+    Args:
+        file_path (str): Filepath of for which the githash should be retrieved.
+
+    Returns:
+        (str): Full 40char length githash of the latest commit this file was changed.
+
+        Example:
+                3b3397ebc2777f47b1ae5258afc4d738095adb83
+    """
+    try:
+        abs_path = Path(file_path).resolve()
+        # HACK: could not think of a better way to get the root directory.
+        score_path = abs_path.__str__().split("score")[0] + "score"
```

Comment:
We have different approaches throughout the code base. We should align how we do that.

Proposal: merge as-is, and address later.

Reply:
Note: here you can also simply use `pathlib.Path(abs_path).parent` as you don't really need the git root.